src/iterators/bfs.jl

```diff
@@ -0,0 +1,95 @@
+"""
+    BFSIterator
+
+`BFSIterator` is a subtype of [`VertexIterator`](@ref) to iterate through graph vertices using a breadth-first search. A source node(s) is optionally supplied as an `Int` or an array-like type that can be indexed if supplying multiple sources. If no source is provided, it defaults to the first vertex.
+    
+# Examples
+```julia-repl
+julia> g = smallgraph(:house)
+{5, 6} undirected simple Int64 graph
+
+julia> for node in BFSIterator(g,3)
+           display(node)
+       end
+3
+1
+4
+5
+2
+```
+"""
+struct BFSIterator{S,G<:AbstractGraph}
+    graph::G
+    source::S
+end
+
+"""
+    BFSVertexIteratorState
+
+`BFSVertexIteratorState` is a struct to hold the current state of iteration
+in BFS which is needed for the `Base.iterate()` function.
+"""
+mutable struct BFSVertexIteratorState
+    visited::BitVector
+    queue::Vector{Int}
```

Comment:
Something from https://github.com/JuliaCollections/DataStructures.jl might be more appropriate

Reply:
I don't think so, removing from the start and the end is O(1) in a Vector if I'm not mistaken

Reply:
I'm assuming the devil is in the constants, otherwise we wouldn't need stacks and queues

Reply:
will look for it, but I still bet on Vectors :D

Reply:
a fair assumption ^^

Reply:
checked...I lost the bet :D

Reply:
woops ^^

Reply:
actually I checked before only on some micro-benchmarks but then I tried in the specific case on a simple graph and I actually saw a regression of 40% so in the end I think we can keep Vectors for now

Reply:
since this feature does not exist yet, it's better to have it sooner than to have a fully optimized version maybe never

Reply:
so let's do it this way and we can improve later if necessary